alchemiscale/storage/statestore.py

```diff
@@ -1452,56 +1458,252 @@ def get_task_failures(self, task: ScopedKey) -> List[ProtocolDAGResultRef]:
+
+        As per the design of the `Task` data lifecycle `waiting`, `running` and `complete`
+        tasks can be set to `error`.
+        """
+        if isinstance(task, ScopedKey):
+            task = [task]
+
+        with self.transaction() as tx:
+            for t in task:
+                q = f"""
+                MATCH (t:Task {{_scoped_key: '{t}'}})
+                RETURN t
+                """
+                task = tx.run(q).to_subgraph()
+                status = task.get("status")
+                if status == TaskStatusEnum.error.value:
+                    continue  # no-op
+                if (
+                    (status != TaskStatusEnum.running.value)
+                    and (status != TaskStatusEnum.waiting.value)
+                    and (status != TaskStatusEnum.complete.value)
+                ):
+                    raise ValueError(
+                        f"Cannot set task {t} with current status: {status} to `error` as it is not currently `running`, `waiting` or `complete`"
+                    )
+                q2 = f"""
+                MATCH (t:Task {{_scoped_key: '{t}'}})
+                SET t.status = '{TaskStatusEnum.error.value}'
+                """
+                tx.run(q2)
 
     def set_task_invalid(
         self,
-        task: Union[Task, ScopedKey],
-        cancel=True,
-    ):
-        ...
+        task: Union[ScopedKey, List[ScopedKey]],
+    ) -> None:
+        """
+        Set the status of a task or list of tasks to `invalid`.
+
+        As per the design of the `Task` data lifecycle any task can be set to `invalid`.
+        Once `invalid` is set the task cannot change status
+        """
+        if isinstance(task, ScopedKey):
+            task = [task]
+
+        with self.transaction() as tx:
+            for t in task:
+                # set the status and delete the ACTIONS relationship
+                q = f"""
+                MATCH (t:Task {{_scoped_key: '{t}'}})
+                SET t.status = '{TaskStatusEnum.invalid.value}'
+                WITH t
+                OPTIONAL MATCH (t)<-[ar:ACTIONS]-(th:TaskHub)
+                DETACH DELETE ar
+                """
+                tx.run(q)
 
     def set_task_deleted(
         self,
-        task: Union[Task, ScopedKey],
-        cancel=True,
-    ):
-        ...
+        task: Union[ScopedKey, List[ScopedKey]],
+    ) -> None:
+        """
+        Set the status of a task or list of tasks to `deleted`.
+
+        As per the design of the `Task` data lifecycle any task can be set to `deleted`.
+        Once `deleted` is set the task cannot change status
+        """
+        if isinstance(task, ScopedKey):
+            task = [task]
+
+        with self.transaction() as tx:
+            for t in task:
+                # set the status and delete any ACTIONS relationship
+                q = f"""
+                MATCH (t:Task {{_scoped_key: '{t}'}})
+                SET t.status = '{TaskStatusEnum.deleted.value}'
```

Comment:
Can simplify due to string `__repr__` casting

Reply:
This doesn't work as the __repr__ of `TaskStatusEnum.deleted` is `TaskStatusEnum.deleted` rather than `deleted`